libbpf-tools/syscount.c

```diff
@@ -0,0 +1,466 @@
+		return true;
+
+	for ( ;; ) {
+		err = bpf_map_get_next_key(fd, &key, &next_key);
+		if (err && errno != ENOENT) {
+			warn("failed to get next key: %s\n", strerror(errno));
+			return false;
+		} else if (err) {
+			break;
+		}
+
+		key = next_key;
+		err = bpf_map_lookup_elem(fd, &key, &vals[i]);
+		if (err && errno != ENOENT) {
+			warn("failed to lookup element: %s\n", strerror(errno));
+			return false;
+		}
+		vals[i++].key = key;
+
+		err = bpf_map_delete_elem(fd, &key);
+		if (err) {
+			warn("failed to delete element: %s\n", strerror(errno));
+			return false;
+		}
+	}
+
+	*count = i;
+	return true;
+}
+
+static error_t parse_arg(int key, char *arg, struct argp_state *state)
+{
+	int number;
+	int err;
+
+	switch (key) {
+	case 'v':
+		env.verbose = true;
+		break;
+	case 'x':
+		env.failures = true;
+		break;
+	case 'L':
+		env.latency = true;
+		break;
+	case 'm':
+		env.milliseconds = true;
+		break;
+	case 'P':
+		env.process = true;
+		break;
+	case 'p':
+		err = get_int(arg, &env.pid, 1, INT_MAX);
+		if (err) {
+			warn("invalid PID: %s\n", arg);
+			argp_usage(state);
+		}
+		break;
+	case 'i':
+		err = get_int(arg, &env.interval, 0, INT_MAX);
+		if (err) {
+			warn("invalid INTERVAL: %s\n", arg);
+			argp_usage(state);
+		}
+		break;
+	case 'd':
+		err = get_int(arg, &env.duration, 1, INT_MAX);
+		if (err) {
+			warn("invalid DURATION: %s\n", arg);
+			argp_usage(state);
+		}
+		break;
+	case 'T':
+		err = get_int(arg, &env.top, 1, INT_MAX);
+		if (err) {
+			warn("invalid TOP: %s\n", arg);
+			argp_usage(state);
+		}
+		break;
+	case 'e':
+		err = get_int(arg, &number, 1, INT_MAX);
+		if (err) {
+			number = errno_by_name(arg);
+			if (number < 0) {
+				warn("invalid errno: %s (bad, or can't "
+				     "parse dynamically; consider using "
+				     "numeric value and/or installing the "
+				     "errno program from moreutils)\n", arg);
+				argp_usage(state);
+			}
+		}
+		env.filter_errno = number;
+		break;
+	case 'l':
+		env.list_syscalls = true;
+		break;
+	default:
+		return ARGP_ERR_UNKNOWN;
+	}
+	return 0;
+}
+
+static volatile sig_atomic_t hang_on = 1;
+
+void sig_int(int signo)
+{
+	hang_on = 0;
+}
+
+/*
+ * attach_progs() differs from the syscount_bpf__attach() skeleton function in
+ * that it will not attach the "sys_enter" tracepoint when there are no latency
+ * measurements needed.
+ */
+static int attach_progs(struct syscount_bpf *obj)
+{
+	struct bpf_object_skeleton *s = obj->skeleton;
+	struct bpf_program *prog;
+	struct bpf_link **link;
+	const char *title;
+	int err;
+	int i;
+
+	for (i = 0; i < s->prog_cnt; i++) {
+		prog = *s->progs[i].prog;
+		link = s->progs[i].link;
+
+		title = bpf_program__title(prog, false);
+		if (!env.latency && strstr(title, "sys_enter"))
+			continue;
+
+		*link = bpf_program__attach(prog);
+		err = libbpf_get_error(*link);
+		if (err) {
+			warn("failed to auto-attach program '%s'\n",
+			     bpf_program__name(prog));
+			return err;
+		}
+	}
+
+	return 0;
+}
```

Comment:
I don't get why you need this generic function. You have exactly 2 programs, and in some cases you want to attach only one of them. Here's the simplest and most straightforward way to do this:
```
obj->links.sys_exit = bpf_program__attach(obj->progs.sys_exit);
err = libbpf_get_error(obj->links.sys_exit);
if (err) /* ... */
if (env.latency) {
    obj->links.sys_enter = bpf_program__attach(obj->progs.sys_enter);
    err = libbpf_get_error(obj->links.sys_exit);
    if (err) /* ... */
}
```

Instead you are using obj->skeleton and use stuff that's meant for libbpf itself only. For no good reason, actually.

BTW, I'd call sys_enter/sys_exit programs something more short (I used just `sys_enter`/`sys_exit`) in the example above, so that it's easier to use from user-space.

Reply:
> Instead you are using obj->skeleton and use stuff that's meant for libbpf itself only. For no good reason, actually.

You're right, I will fix this.

BTW, I've found a bug which I didn't see before (probably my system wasn't loaded enough, and also I've used batch version of lookup_and_delete initially before you rejected the same thing in vfsstat). There is a race in the `read_vals()`: the loop is not atomic and consists of multiple `bpf` system calls. Thus, the `bpf` entry in the map may be re-populated until it is the last entry in the map, and `bpf` entry could appear multiple times.

In the original tool, it does the printing as follows:
```
items = data.items() // read all values from the map one-by-one
keys = data.keys() // get all map keys
for k in keys: delete(k) // delete one element from the map
```
this also has some problems, because it looses system calls which are listed in keys and happened between getting items and deleting elements by keys.

The right solution which doesn't seem to loose any syscalls would be to do
```
keys = get_keys()
items = get_and_delete_items(keys)
```
but the `bpf_map_lookup_and_delete_elem` is supported only for stacks/queues.

Should I just use the approach used in syscount.py and loose some (relatively few) syscalls? Or maybe the right way is to patch `bpf_map_lookup_and_delete_elem` to support hashes? (Or there is some reason why it wasn't patched when batch lookup_and_delete was added?)

Reply:
Don't know about lookup_and_delete for hashmap, honestly.

I'd just use two hashmaps and switch between them for each reporting period. From user-space, before you start reporting, increment an index ("current map index"), re-routing BPF logic to a different map. Then work with the one that's not modified, read it, clear it, then on next reporting period bump counter again. On BPF side, before updating data map, you just:
```
int cur_map_idx = 0; /* this gets updated from user-space */

SEC(...)
int prog(...) {
    void *data = (cur_map_idx & 1) ? data1 : data2;
    /* proceed as before */
}
```

There is still tiny race (after you bumped cur_map_idx, there could be still BPF programs that are still updating previous map), but I think they are so tiny that it doesn't matter in practice.

BTW, to mitigate some of the memory overhead, I think it should be fine to use BPF_F_NO_PREALLOC hashmap (just specify it as map_flags). It will make kernel allocate memory for hash map entries on demand. So even with big hashmap, if it doesn't have lots of entries, you'd still have a manageable memory usage.

Makes sense?

Reply:
So, if we still have a race, then I, probably, would prefer to read data the same way the original `syscount.py` — in this way I can re-use functions to read keys and hash items later. But I will try both variants first and see what looks better.

Reply:
> BTW, to mitigate some of the memory overhead, I think it should be fine to use BPF_F_NO_PREALLOC hashmap (just specify it as map_flags). It will make kernel allocate memory for hash map entries on demand. So even with big hashmap, if it doesn't have lots of entries, you'd still have a manageable memory usage.

Thanks, will add this.